src/Voron/Data/CompactTrees/CompactTree.cs

```diff
@@ -549,19 +547,70 @@ private void MaybeMergeEntries(ref CursorState destinationState)
 
             var srcDictionary = GetEncodingDictionary(sourceHeader->DictionaryId);
             var destDictionary = GetEncodingDictionary(destinationHeader->DictionaryId);
-            bool reencode = sourceHeader->DictionaryId != destinationHeader->DictionaryId;
+            bool reEncode = sourceHeader->DictionaryId != destinationHeader->DictionaryId;
 
-            int sourceEncodedKeysLenght = 0;
+            int sourceEncodedKeysLength = 0;
             int sourceKeysCopied = 0;
-            for (; sourceKeysCopied < sourceHeader->NumberOfEntries; sourceKeysCopied++)
+            if (reEncode) // avoid branch inside loop by duplicating the loop
+            {
+                for (; sourceKeysCopied < sourceHeader->NumberOfEntries; sourceKeysCopied++)
+                {
+                    if (MoveEntryWithReEncoding(decodeBuffer, encodeBuffer, ref destinationState, entries) == false)
```

Comment:
Do you have a profile that shows a material difference here? Because, given that we are executing for many (hundreds) of iterations every time, the branch detector will take care of it. If there is a difference, I will look into what it triggering the JIT to underperform there, because there should be no differences.

Reply:
Done